nativelib/src/main/scala-next/scala/scalanative/safe/SafeZone.scala

```diff
@@ -0,0 +1,51 @@
+package scala.scalanative.safe
+
+import language.experimental.captureChecking
+import scalanative.unsigned._
+import scala.annotation.implicitNotFound
+import scala.scalanative.runtime.{RawPtr, CZone}
+
+@implicitNotFound("Given method requires an implicit zone.")
+trait SafeZone {
+
+  /** Return this zone allocator is open or not. */
+  def isOpen: Boolean
+
+  /** Return this zone allocator is closed or not. */
+  def isClosed: Boolean = !isOpen
+
+  /** Require this zone allocator is open. */
+  def checkOpen(): Unit = {
+    if (!isOpen)
+      throw new IllegalStateException(s"Zone ${this} is already closed.")
+  }
+
+  /** Frees allocations. This zone allocator is not reusable once closed. */
+  private[scalanative] def close(): Unit
+
+  /** Return the handle of this zone allocator. */
+  private[scalanative] def handle: RawPtr
+}
+
+final class MemorySafeZone (private[scalanative] val handle: RawPtr) extends SafeZone {
+
+  private[this] var flagIsOpen = true
+
+  override def close(): Unit = {
+    checkOpen()
+    flagIsOpen = false
+    CZone.close(handle)
+  }
+
+  override def isOpen: Boolean = flagIsOpen
+}
+
+
+object SafeZone {
+  /** Run given function with a fresh zone and destroy it afterwards. */
+  final def apply[T](f: ({*} SafeZone) => T): T = {
```

Comment:
I think context function would make it way more user friendly in the long run. In most of the cases we only want to allocate in the current Zone scope. It would match perfectly with the additional alloc method defined lower
```suggestion
  final def apply[T](f: ({*} SafeZone) ?=> T): T = {
```
here's how users code can then look like: 
```scala
  SafeZone:
    val str = alloc(String("fooBar"))
    val zone = summon[SafeZone]
    val str2= zone.alloc(String("foo"))
    val str3 = summon[SafeZone] alloc String("baz")

    SafeZone{
      val str2 = alloc(new String("bar"))
    }
    ```

Reply:
you could even add
```scala
transparent inline def zone(using sz: {*} SafeZone): {sz} SafeZone = sz
```
then the zone can be summoned more elegantly

Reply:
Changing `apply` to context function will disallow users to explicit declare the SafeZone value, i.e. write code like `SafeZone { sz => ...}`, right? I don't think it worths.

Reply:
you can still have `SafeZone { sz ?=> ...}`

Reply:
> you could even add
> 
> ```scala
> transparent inline def zone(using sz: {*} SafeZone): {sz} SafeZone = sz
> ```
> 
> then the zone can be summoned more elegantly

Sorry I didn't get it, what's the benefit we get from calling such a `zone` function that we can't get from calling `summon[SafeZone]`?

Reply:
would you rather type `summon[Context]` in the dotty compiler or type `ctx` :) because its the same thing